### ISSUE TITLE
Bug: `Array to string conversion` in resolver for `term` query on `PostObject`

### ISSUE THREAD

**Issue by hews** [wp-graphql/wp-graphql#400]
WP version: 4.9.1   
WP-GraphQL version: 0.0.25   
OS: Debian Jessie

I am attempting to query all terms connected to a post:

```
query {
    postBy( id: 1 ) {
        postId
        title

        terms( taxonomy: CATEGORY ) {
            ... on Category {
                name
            }
        }

        termNames( taxonomy: CATEGORY )
    }
}
```

In this instance, I am receiving the errors: `Array to string conversion.` I could give you the stack trace, but it's entirely unhelpful as the error is coming from a resolver; the operative lines are:

```
...
PHP  38. get_the_terms($post = *uninitialized*, $taxonomy = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/src/Type/PostObject/PostObjectType.php:386
...
PHP  38. get_the_terms($post = *uninitialized*, $taxonomy = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/src/Type/PostObject/PostObjectType.php:416
```

It seems that, in lieu of using the list of `$allowed_taxonomies` registered in the plugin, since I am passing in a taxonomy as an arg, I am using this logic:

```
# src/Type/PostObject/PostObjectType.php#L383
$allowed_taxonomies = ! empty( $args['taxonomy'] ) ? [ $args['taxonomy'] ] : $allowed_taxonomies;
```

The problem being that the argument is *itself* an array, ie:

```php
print_r( $args );
# Array
# (
#     [taxonomy] => Array
#         (
#             [0] => category
#         )
# 
# )

$allowed_taxonomies = ! empty( $args['taxonomy'] ) ? [ $args['taxonomy'] ] : $allowed_taxonomies;

print_r( $allowed_taxonomies );
# Array
# (
#     [0] => Array
#         (
#             [0] => category
#         )
# 
# )
```

Then, when we iterate on `$allowed_taxonomies`, and try to use the first item, we are using an array instead of a string:

```
# src/Type/PostObject/PostObjectType.php#L385
foreach ( $allowed_taxonomies as $taxonomy ) {
    $tax_terms = get_the_terms( $post->ID, $taxonomy ); # $taxonomy is an array
    ...
```

By making the changes below, I removed the errors:

```
# src/Type/PostObject/PostObjectType.php#L383
- $allowed_taxonomies = ! empty( $args['taxonomy'] ) ? [ $args['taxonomy'] ] : $allowed_taxonomies;
+ $allowed_taxonomies = ! empty( $args['taxonomy'] ) ? $args['taxonomy'] : $allowed_taxonomies;

# src/Type/PostObject/PostObjectType.php#L413
- $allowed_taxonomies = ! empty( $args['taxonomy'] ) ? [ $args['taxonomy'] ] : $allowed_taxonomies;
+ $allowed_taxonomies = ! empty( $args['taxonomy'] ) ? $args['taxonomy'] : $allowed_taxonomies;
```

I can submit a PR for this fix.

### When reporting a _bug_, please be sure to include the following:

- [x] A descriptive title
- [x] An *isolated* way to reproduce the behavior (example: GitHub repository with code isolated to the issue that anyone can clone to observe the problem)
- [x] What version of `WPGraphQL` you're using, and the platform(s) you're running it on
- [x] What other plugins you're using (n/a)
- [x] The behavior you expect to see, and the actual behavior

**Comment by jasonbahl** [comment]
@hews this looks like it breaks things now. . .see: https://travis-ci.org/wp-graphql/wp-graphql/builds/341673493?utm_source=github_status&utm_medium=notification

We need the allowed taxonomies to be an array, so the current behavior is correct. 

I did, however, come across an issue with the use of `get_the_terms()` inside that resolver, so perhaps that's what you were experiencing. 

I'll open a Pull Request with my changes, and let's see if that resolves the issue you were running into.

**Comment by hews** [comment]
One full stack trace. Note: line numbers are off bc I have some debug statements in there.

```
14-Feb-2018 22:26:10 UTC] PHP Notice:  Array to string conversion in /var/www/web/wp/wp-includes/taxonomy.php on line 3102
[14-Feb-2018 22:26:10 UTC] PHP Stack trace:
[14-Feb-2018 22:26:10 UTC] PHP   1. {main}() /var/www/web/index.php:0
[14-Feb-2018 22:26:10 UTC] PHP   2. require() /var/www/web/index.php:5
[14-Feb-2018 22:26:10 UTC] PHP   3. wp($query_vars = *uninitialized*) /var/www/web/wp/wp-blog-header.php:16
[14-Feb-2018 22:26:10 UTC] PHP   4. WP->main($query_args = *uninitialized*) /var/www/web/wp/wp-includes/functions.php:960
[14-Feb-2018 22:26:10 UTC] PHP   5. WP->parse_request($extra_query_vars = *uninitialized*) /var/www/web/wp/wp-includes/class-wp.php:713
[14-Feb-2018 22:26:10 UTC] PHP   6. do_action_ref_array($tag = *uninitialized*, $args = *uninitialized*) /var/www/web/wp/wp-includes/class-wp.php:373
[14-Feb-2018 22:26:10 UTC] PHP   7. WP_Hook->do_action($args = *uninitialized*) /var/www/web/wp/wp-includes/plugin.php:515
[14-Feb-2018 22:26:10 UTC] PHP   8. WP_Hook->apply_filters($value = *uninitialized*, $args = *uninitialized*) /var/www/web/wp/wp-includes/class-wp-hook.php:310
[14-Feb-2018 22:26:10 UTC] PHP   9. WPGraphQL\Router->resolve_http_request(*uninitialized*) /var/www/web/wp/wp-includes/class-wp-hook.php:286
[14-Feb-2018 22:26:10 UTC] PHP  10. WPGraphQL\Router::process_http_request() /var/www/web/app/mu-plugins/wp-graphql/src/Router.php:147
[14-Feb-2018 22:26:10 UTC] PHP  11. GraphQL\Server\StandardServer->executeRequest($parsedBody = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/src/Router.php:437
[14-Feb-2018 22:26:10 UTC] PHP  12. GraphQL\Server\Helper->executeOperation($config = *uninitialized*, $op = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Server/StandardServer.php:130
[14-Feb-2018 22:26:10 UTC] PHP  13. GraphQL\Server\Helper->promiseToExecuteOperation($promiseAdapter = *uninitialized*, $config = *uninitialized*, $op = *uninitialized*, $isBatch = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Server/Helper.php:176
[14-Feb-2018 22:26:10 UTC] PHP  14. GraphQL\GraphQL::promiseToExecute($promiseAdapter = *uninitialized*, $schema = *uninitialized*, $source = *uninitialized*, $rootValue = *uninitialized*, $context = *uninitialized*, $variableValues = *uninitialized*, $operationName = *uninitialized*, $fieldResolver = *uninitialized*, $validationRules = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Server/Helper.php:260
[14-Feb-2018 22:26:10 UTC] PHP  15. GraphQL\Executor\Executor::promiseToExecute($promiseAdapter = *uninitialized*, $schema = *uninitialized*, $ast = *uninitialized*, $rootValue = *uninitialized*, $contextValue = *uninitialized*, $variableValues = *uninitialized*, $operationName = *uninitialized*, $fieldResolver = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/GraphQL.php:155
[14-Feb-2018 22:26:10 UTC] PHP  16. GraphQL\Executor\Executor->doExecute() /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Executor.php:151
[14-Feb-2018 22:26:10 UTC] PHP  17. GraphQL\Executor\Promise\Adapter\SyncPromiseAdapter->create($resolver = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Executor.php:288
[14-Feb-2018 22:26:10 UTC] PHP  18. GraphQL\Executor\Executor->GraphQL\Executor\{closure}($resolve = *uninitialized*, *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Promise/Adapter/SyncPromiseAdapter.php:59
[14-Feb-2018 22:26:10 UTC] PHP  19. GraphQL\Executor\Executor->executeOperation($operation = *uninitialized*, $rootValue = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Executor.php:287
[14-Feb-2018 22:26:10 UTC] PHP  20. GraphQL\Executor\Executor->executeFields($parentType = *uninitialized*, $source = *uninitialized*, $path = *uninitialized*, $fields = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Executor.php:324
[14-Feb-2018 22:26:10 UTC] PHP  21. GraphQL\Executor\Executor->resolveField($parentType = *uninitialized*, $source = *uninitialized*, $fieldNodes = *uninitialized*, $path = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Executor.php:442
[14-Feb-2018 22:26:10 UTC] PHP  22. GraphQL\Executor\Executor->completeValueCatchingError($returnType = *uninitialized*, $fieldNodes = *uninitialized*, $info = *uninitialized*, $path = *uninitialized*, $result = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Executor.php:721
[14-Feb-2018 22:26:10 UTC] PHP  23. GraphQL\Executor\Executor->completeValueWithLocatedError($returnType = *uninitialized*, $fieldNodes = *uninitialized*, $info = *uninitialized*, $path = *uninitialized*, $result = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Executor.php:799
[14-Feb-2018 22:26:10 UTC] PHP  24. GraphQL\Executor\Executor->completeValue($returnType = *uninitialized*, $fieldNodes = *uninitialized*, $info = *uninitialized*, $path = *uninitialized*, $result = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Executor.php:845
[14-Feb-2018 22:26:10 UTC] PHP  25. GraphQL\Executor\Executor->completeListValue($returnType = *uninitialized*, $fieldNodes = *uninitialized*, $info = *uninitialized*, $path = *uninitialized*, $result = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Executor.php:937
[14-Feb-2018 22:26:10 UTC] PHP  26. GraphQL\Executor\Executor->completeValueCatchingError($returnType = *uninitialized*, $fieldNodes = *uninitialized*, $info = *uninitialized*, $path = *uninitialized*, $result = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Executor.php:1175
[14-Feb-2018 22:26:10 UTC] PHP  27. GraphQL\Executor\Executor->completeValueWithLocatedError($returnType = *uninitialized*, $fieldNodes = *uninitialized*, $info = *uninitialized*, $path = *uninitialized*, $result = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Executor.php:799
[14-Feb-2018 22:26:10 UTC] PHP  28. GraphQL\Executor\Executor->completeValue($returnType = *uninitialized*, $fieldNodes = *uninitialized*, $info = *uninitialized*, $path = *uninitialized*, $result = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Executor.php:845
[14-Feb-2018 22:26:10 UTC] PHP  29. GraphQL\Executor\Executor->completeAbstractValue($returnType = *uninitialized*, $fieldNodes = *uninitialized*, $info = *uninitialized*, $path = *uninitialized*, $result = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Executor.php:961
[14-Feb-2018 22:26:10 UTC] PHP  30. GraphQL\Executor\Executor->completeObjectValue($returnType = *uninitialized*, $fieldNodes = *uninitialized*, $info = *uninitialized*, $path = *uninitialized*, $result = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Executor.php:1098
[14-Feb-2018 22:26:10 UTC] PHP  31. GraphQL\Executor\Executor->collectAndExecuteSubfields($returnType = *uninitialized*, $fieldNodes = *uninitialized*, $info = *uninitialized*, $path = *uninitialized*, $result = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Executor.php:1250
[14-Feb-2018 22:26:10 UTC] PHP  32. GraphQL\Executor\Executor->executeFields($parentType = *uninitialized*, $source = *uninitialized*, $path = *uninitialized*, $fields = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Executor.php:1305
[14-Feb-2018 22:26:10 UTC] PHP  33. GraphQL\Executor\Executor->resolveField($parentType = *uninitialized*, $source = *uninitialized*, $fieldNodes = *uninitialized*, $path = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Executor.php:442
[14-Feb-2018 22:26:10 UTC] PHP  34. GraphQL\Executor\Executor->resolveOrError($fieldDef = *uninitialized*, $fieldNode = *uninitialized*, $resolveFn = *uninitialized*, $source = *uninitialized*, $context = *uninitialized*, $info = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Executor.php:713
[14-Feb-2018 22:26:10 UTC] PHP  35. WPGraphQL\Utils\InstrumentSchema::WPGraphQL\Utils\{closure}($source = *uninitialized*, $args = *uninitialized*, $context = *uninitialized*, $info = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Executor.php:750
[14-Feb-2018 22:26:10 UTC] PHP  36. call_user_func:{/var/www/web/app/mu-plugins/wp-graphql/src/Utils/InstrumentSchema.php:133}(*uninitialized*, *uninitialized*, *uninitialized*, *uninitialized*, *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/src/Utils/InstrumentSchema.php:133
[14-Feb-2018 22:26:10 UTC] PHP  37. WPGraphQL\Type\PostObject\PostObjectType::WPGraphQL\Type\PostObject\{closure}($post = *uninitialized*, $args = *uninitialized*, $context = *uninitialized*, $info = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/src/Utils/InstrumentSchema.php:133
[14-Feb-2018 22:26:10 UTC] PHP  38. get_the_terms($post = *uninitialized*, $taxonomy = *uninitialized*) /var/www/web/app/mu-plugins/wp-graphql/src/Type/PostObject/PostObjectType.php:392
[14-Feb-2018 22:26:10 UTC] PHP  39. get_object_term_cache($id = *uninitialized*, $taxonomy = *uninitialized*) /var/www/web/wp/wp-includes/category-template.php:1195
```

**Comment by hews** [comment]
> We need the allowed taxonomies to be an array, so the current behavior is correct.

I think you aren't hearing what I'm laying down. `$args['taxonomy']` ***is an array*** when I am using it, and then you are wrapping it in *another array* with `$allowed_taxonomies= ... [ $args['taxonomy'] ]`.

I don't really see how this could be dependent on my environment, since `$args['taxonomies']` is defined to be a list of `TaxonomyEnum`s.

**Comment by hews** [comment]
Might this be some lacunae in your tests?

For easy reference:

https://github.com/wp-graphql/wp-graphql/blob/daa18778829c7f92fec1010bef4785f03b3bb43f/tests/wpunit/PostObjectQueriesTest.php#L436-L442

You're using one test each for `termNames` and `terms`. For `terms` you are not passing any args: this behavior is not causing me errors either.

However, for termNames you are using a syntax I am not aware of:

```gql
    ...
    tagNames:termNames(taxonomy:TAG)
    ...
```

... where I am using:

```gql
    ...
    termNames(taxonomy:TAG)
    ...
```

... which is causing me error.

**Comment by jasonbahl** [comment]
closed by #401
